### ISSUE TITLE
Problems using ErrorThroughSurface

### ISSUE THREAD

**Issue by ElizaJayne11** [OceanParcels/parcels#699]
Hi
I have been trying to implement a simulation where particles swim toward the surface. I want to ensure that they do not go out the surface but are instead confined to the top (i.e. not deleted), and only delete particles that are transported laterally out of the domain. 
I had seen your recommendations to use a recovery kernel with ErrorCode.ErrorThroughSurface, and I have tried implementing that along with a version of your SubmergedParticle kernel. Pieces of my code are shown below:

from parcels import ErrorCode

def SubmergeParticle(particle, fieldset, time):
        particle.depth = 0.51 #Minimum depth os 0.5m
        particle.time = time + particle.dt  # to not trigger kernels again, otherwise infinite loop       

def DeleteParticle(particle, fieldset, time):
    particle.delete()

pset.execute(kernels, runtime=delta(days=3), 
                      dt=delta(hours=1),   
                      output_file=pfile, 
                      recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle,
                                ErrorCode.ErrorThroughSurface: SubmergeParticle})

However, I keep getting the error: 

  File "/Applications/anaconda3/envs/py3_parcels/lib/python3.7/enum.py", line 349, in __getattr__
    raise AttributeError(name) from None

AttributeError: ErrorThroughSurface

I have tried several things to make this work, including importing ErrorCode from parcels.tools.error (I noticed that parcels.kernel.error does not reference ErrorThroughSurface), and running just the ErrorThroughSurface recovery option. I still get the same error. Although, when I just have ErrorOutOf Bounds the code will run, but of course, this does not afford the functionality I am seeking.

Any advice would be most welcome.

**Comment by erikvansebille** [comment]
Dear @ElizaJayne11, I'm happy to help but could you  provide the _full_ error code you got? That will significantly help with debugging

**Comment by ElizaJayne11** [comment]
Hi Erik
Wow! You are very quick to respond. Thanks again for your help. Below is my full error code.
Traceback (most recent call last):

  File "<ipython-input-1-5e23f545d2f1>", line 1, in <module>
    runfile('/Users/wepowerthestars/Desktop/CONNECTIVITY/Scallops 3D/Variable Depth ICs/No Buried ICs 2.py', wdir='/Users/wepowerthestars/Desktop/CONNECTIVITY/Scallops 3D/Variable Depth ICs')

  File "/Applications/anaconda3/envs/py3_parcels/lib/python3.7/site-packages/spyder_kernels/customize/spydercustomize.py", line 827, in runfile
    execfile(filename, namespace)

  File "/Applications/anaconda3/envs/py3_parcels/lib/python3.7/site-packages/spyder_kernels/customize/spydercustomize.py", line 110, in execfile
    exec(compile(f.read(), filename, 'exec'), namespace)

  File "/Users/wepowerthestars/Desktop/CONNECTIVITY/Scallops 3D/Variable Depth ICs/No Buried ICs 2.py", line 113, in <module>
    ErrorCode.ErrorThroughSurface: SubmergeParticle})

  File "/Applications/anaconda3/envs/py3_parcels/lib/python3.7/enum.py", line 349, in __getattr__
    raise AttributeError(name) from None

AttributeError: ErrorThroughSurface

**Comment by erikvansebille** [comment]
Hmm, I still can't reproduce your bug. Could you see if the code below gives the same bug for you? And if not, could you explore a bit what you need to change in the code below in order for the bug to appear? That would really make it much easier to debug for us!

```python
from parcels import ErrorCode, FieldSet, AdvectionRK4_3D, ParticleSet, JITParticle
import numpy as np

def SubmergeParticle(particle, fieldset, time):
    particle.depth = 0.51  # Minimum depth is 0.5m
    particle.time = time + particle.dt  # to not trigger kernels again, otherwise infinite loop

def DeleteParticle(particle, fieldset, time):
    particle.delete()

xdim = ydim = zdim = 2
dimensions = {'lon': np.linspace(0., 1, xdim, dtype=np.float32),
              'lat': np.linspace(0., 1, ydim, dtype=np.float32),
              'depth': np.linspace(0., 1, zdim, dtype=np.float32)}
data = {'U': 0.01*np.ones((xdim, ydim, zdim), dtype=np.float32),
        'V': np.zeros((xdim, ydim, zdim), dtype=np.float32),
        'W': -1 * np.ones((xdim, ydim, zdim), dtype=np.float32)}
fieldset = FieldSet.from_data(data, dimensions, mesh='flat')

pset = ParticleSet(fieldset=fieldset, pclass=JITParticle,
                   lon=0.5, lat=0.5, depth=0.9)

pset.execute(AdvectionRK4_3D, runtime=10, dt=1,
             recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle,
                       ErrorCode.ErrorThroughSurface: SubmergeParticle})
```

**Comment by ElizaJayne11** [comment]
Hi
Success and embarrassment....I tried running your code and got the same error, which made me suspicious that the problem had to do with my parcels code. I checked my installed files and discovered that I was using error.py from July 2, which did not have an ErrorThroughSurface recovery option. I updated the error.py to the latest version whereupon both your code and my application now compile fine. 
I apologize for taking up your time on this. I am still learning all about Parcels, Python and GitHub and finding it a steep learning curve. Thanks for your support.
